### PR TITLE
RR-317 - Database, entity and mappers for persisting Timeline events

### PR DESCRIPTION
**Pull request by mattwills-moj-digital**
This PR introduces the database tables for a Prisoner's Timeline, along with the required JPA entities and repository.

It's larger than anticipated, because I had to make changes to `GoalService` and `JpaGoalPersistenceAdapter` to ensure that the correct events are created whenever an ActionPlan is created, when adding a `Goal`.

I'm going to add the REST endpoint to retrieve a Prisoner's Timeline, along with the required integration tests, in a separate PR.

### CODE REVIEW

File: domain/goal/src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/GoalService.kt [hunk 1]
```diff
@@ -21,19 +22,27 @@ private val log = KotlinLogging.logger {}
...
     log.info { "Creating new Goal for prisoner [$prisonNumber]" }
-    return persistenceAdapter.createGoal(prisonNumber, createGoalDto)
-      .also {
-        goalEventService.goalCreated(prisonNumber, it)
-      }
+
+    // TODO RR-227 - We need to change throw a 404 once the create action plan endpoint is being called by the UI (with an optional review date)
+    return if (actionPlanDoesNotExist(prisonNumber)) {
+      actionPlanPersistenceAdapter.createActionPlan(newActionPlan(prisonNumber, createGoalDto))
+        .also { actionPlanEventService.actionPlanCreated(it) }
```
Comment (mattwills-moj-digital) [inline domain/goal/src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/GoalService.kt:40]: This ensures an `ACTION_PLAN_CREATED` `TimelineEvent` is stored.

File: domain/timeline/src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/timeline/Timeline.kt [hunk 2]
```diff
@@ -1,10 +1,13 @@
 package uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline
 
+import java.util.UUID
+
 /**
  * A Timeline is sequenced list of [TimelineEvent]s for a given prisoner. The list can be empty, representing an empty
  * timeline for the prisoner.
  */
 class Timeline(
+  val reference: UUID,
```
Comment (mattwills-moj-digital) [inline domain/timeline/src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/timeline/Timeline.kt:10]: Not strictly needed, but I think it's best if every resource has a reference.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaTimelinePersistenceAdapter.kt [hunk 3]
```diff
@@ -3,33 +3,68 @@ package uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa
...
   }
 
   @Transactional
-  override fun recordTimelineEvents(prisonNumber: String, events: List<TimelineEvent>) {
+  override fun recordTimelineEvents(prisonNumber: String, events: List<TimelineEvent>): Timeline {
     log.info { "Recording [${events.size}] TimelineEvents for prisoner [$prisonNumber]" }
-    // TODO RR-317
+
+    val timelineEntity = findOrCreateTimelineEntity(prisonNumber)
+    events.forEach {
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaTimelinePersistenceAdapter.kt:46]: This further addresses [a comment](https://github.com/ministryofjustice/hmpps-education-and-work-plan-api/pull/91#discussion_r1324201100) @nathanrussell-moj-digital made on my previous PR re where this loop lives.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/ActionPlanEntity.kt [hunk 4]
```diff
@@ -71,9 +71,9 @@ class ActionPlanEntity(
     /**
      * Returns new un-persisted [ActionPlanEntity] for the specified prisoner with an empty collection of [GoalEntity]s
      */
-    fun newActionPlanForPrisoner(reference: UUID, prisonNumber: String, reviewDate: LocalDate?): ActionPlanEntity =
+    fun newActionPlanForPrisoner(prisonNumber: String, reviewDate: LocalDate?): ActionPlanEntity =
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/ActionPlanEntity.kt:74]: Unrelated change - just making it consistent with the equivalent convenience method in `TimelineEntity`.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEventEntity.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  /**
+   * The ID of the prison that the event relates to.
+   */
+  @Column(updatable = false)
+  @field:NotNull
+  var createdAtPrison: String? = null,
+
+  /**
+   * Username of the person who created the original event (not to be confused with who/what saved this entity)
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEventEntity.kt:49]: I've added comments to some of these fields, as they're easily confused with fields of the same name in other entities. They're effectively the same, but the intention is different and they're populated manually from the original action/entity, rather than via JPA annotations, such as `@CreatedBy`. I could consider renaming them.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEntity.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  var prisonNumber: String,
+
+  @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
+  @JoinColumn(name = "timeline_id", nullable = false)
+  @OrderBy(value = "createdAt")
+  @field:NotNull
+  var events: MutableList<TimelineEventEntity>? = null,
+
+  @Column(updatable = false)
+  @CreationTimestamp
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEntity.kt:44]: Do these annotations work? (created/updated timestamps)
In the other entity classes we needed to annotate the class with `@EntityListeners(AuditingEntityListener::class)` - but that might have been to get the spring managed fields populated (created/updated by)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/AsyncActionPlanEventService.kt [hunk 7]
```diff
@@ -16,14 +16,22 @@ private val log = KotlinLogging.logger {}
...
+  private val telemetryService: TelemetryService,
   private val timelineEventFactory: TimelineEventFactory,
   private val timelineService: TimelineService,
 ) : ActionPlanEventService {
   override fun actionPlanCreated(actionPlan: ActionPlan) {
     runBlocking {
       log.info { "ActionPlan created event for prisoner [${actionPlan.prisonNumber}]" }
-      val timelineEvent = timelineEventFactory.actionPlanCreatedEvent(actionPlan)
-      launch { timelineService.recordTimelineEvent(actionPlan.prisonNumber, timelineEvent) }
+      launch {
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/AsyncActionPlanEventService.kt:26]: Minor, and it probably doesnt add much (if anything), but I wonder if the launch could/should be in the loop?
```
      actionPlan.goals.forEach {
        launch {
          telemetryService.trackGoalCreateEvent(it)
        }
      }
```

There's probably nothing in it, and I dont think either approach would affect the time take for the user (ie. either do track events in a thread, or track each event in its own thread)

Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/AsyncActionPlanEventService.kt:26]: Ignore me - its probably a premature optimisation

**Review by nathanrussell-moj-digital — DISMISSED** [review]
LGTM 👍 
though I'm not sure about the motivation/rationale for changing the get timeline to return null, rather than a timeline with zero events? 🤔 
I'm sure you had your reasons 😁 , we can talk about it later 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEntity.kt (hunk 6 above)
Comment (mattwills-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/entity/TimelineEntity.kt:44]: Yes they are populated, but you've highlighted the need for an assertion for this in the ITs, which I've just added 👍

**Review by nathanrussell-moj-digital — APPROVED** [review]

**Comment by mattwills-moj-digital** [comment]
> LGTM 👍 though I'm not sure about the motivation/rationale for changing the get timeline to return null, rather than a timeline with zero events? 🤔 I'm sure you had your reasons 😁 , we can talk about it later 👍

Actually, I'm not sure I did have my reasons! 🤣 Moreover, I've stated in the swagger spec (PR coming shortly) that a 404 will be returned if the Timeline doesn't exist, so I never intended to return null.

Anyway, it's a good challenge. I've just updated the `TimelineService` to throw `TimelineDoesNotExistException`.

**Review by nathanrussell-moj-digital — APPROVED** [review]
